pkg/sorter/sorter.go

```diff
@@ -0,0 +1,326 @@
+// the field to sort instance types by. json path must start with "$" character (Ex: "$.MemoryInfo.SizeInMiB").
+//
+// sortDirection represents the direction to sort in. Valid options: "ascending", "asc", "descending", "desc".
+func NewSorter(instanceTypes []*instancetypes.Details, sortField string, sortDirection string) (*Sorter, error) {
+	var isDescending bool
+	switch sortDirection {
+	case sortDescending, sortDesc:
+		isDescending = true
+	case sortAscending, sortAsc:
+		isDescending = false
+	default:
+		return nil, fmt.Errorf("invalid sort direction: %s (valid options: %s, %s, %s, %s)", sortDirection, sortAscending, sortAsc, sortDescending, sortDesc)
+	}
+
+	// Create sorterNode objects for each instance type
+	sorters := []*sorterNode{}
+	for _, instanceType := range instanceTypes {
+		newSorter, err := newSorterNode(instanceType, sortField)
+		if err != nil {
+			return nil, fmt.Errorf("error creating sorting node: %v", err)
+		}
+
+		sorters = append(sorters, newSorter)
+	}
+
+	return &Sorter{
+		sorters:      sorters,
+		sortField:    sortField,
+		isDescending: isDescending,
+	}, nil
+}
+
+// newSorterNode creates a new sorterNode object which represents the given instance type
+// and can be used in sorting of instance types based on the given sortField
+func newSorterNode(instanceType *instancetypes.Details, sortField string) (*sorterNode, error) {
+	// some important fields (such as gpu count) can not be accessed directly in the instancetypes.Details
+	// struct, so we have special hard-coded flags to handle such cases
+	switch sortField {
+	case gpuCountField:
+		gpuCount := getTotalGpusCount(instanceType)
+		return &sorterNode{
+			instanceType: instanceType,
+			fieldValue:   reflect.ValueOf(gpuCount),
+		}, nil
+	case inferenceAcceleratorsField:
+		acceleratorsCount := getTotalAcceleratorsCount(instanceType)
+		return &sorterNode{
+			instanceType: instanceType,
+			fieldValue:   reflect.ValueOf(acceleratorsCount),
+		}, nil
+	}
+
+	// convert instance type into json
+	jsonInstanceType, err := json.Marshal(instanceType)
+	if err != nil {
+		return nil, err
+	}
+
+	// unmarshal json instance types in order to get proper format
+	// for json path parsing
+	var jsonData interface{}
+	err = json.Unmarshal(jsonInstanceType, &jsonData)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the desired field from the json data based on the passed in
+	// json path
+	result, err := jsonpath.JsonPathLookup(jsonData, sortField)
+	if err != nil {
+		// handle case where parent objects in path are null
+		// by setting result to nil
+		if err.Error() == "get attribute from null object" {
+			result = nil
+		} else {
+			return nil, fmt.Errorf("error during json path lookup: %v", err)
+		}
+	}
+
+	return &sorterNode{
+		instanceType: instanceType,
+		fieldValue:   reflect.ValueOf(result),
+	}, nil
+}
+
+// Sort the instance types in the Sorter based on the Sorter's sort field and
+// direction
+func (s *Sorter) Sort() error {
```

Comment:
should be able to use `sort.Slice` instead so you don't have to implement the whole sort interface.

Reply:
That is a good point. I have changed my sorting implementation to use `sort.Slice`